Add chatbot analytics

guarani opened the pull request:
Fixes #21279

## To test

To start testing, first, go to the Help screen. Also, ensure the Contact Support via DocsBot feature flag is ON under App Settings > Debug.

### User chats with the bot
1. Tap "Contact support"
2. Chat with the bot
3. Close the chat by pressing the back button
4. Expect to see the following Tracks events (in the Xcode console or in the app logs):

<details>
<summary>Tracks events (click to expand)</summary>

```
🔵 Tracked: support_opened <>
🔵 Tracked: support_chatbot_started <chat_id: chat UUID>
🔵 Tracked: support_chatbot_ended <chat_id: chat UUID>
```
</details>

### User escalates to ticket
1. Tap "Contact support"
2. Chat with the bot
3. Tap one of the "Contact support" buttons and create a ticket
4. Expect to see the following Tracks events 

<details>
<summary>Tracks events (click to expand)</summary>

```
🔵 Tracked: support_opened <>
🔵 Tracked: support_chatbot_started <chat_id: chat UUID>
🔵 Tracked: support_new_request_created <>
🔵 Tracked: support_chatbot_ticket_success <chat_id: chat UUID>
🔵 Tracked: support_ticket_list_viewed <>
🔵 Tracked: support_chatbot_ended <chat_id: chat UUID>
```
</details>

5. Tap the ticket to view it
6. Expect to see the following Tracks event:
```
 🔵 Tracked: support_ticket_user_viewed <>
```

### Ticket creation fails
1. Tap "Contact support"
2. Chat with the bot
3. Turn off the network to force an error
4. Tap one of the "Contact support" buttons and create a ticket 
5. Expect the ticket creation to fail
6. Expect to see the following Tracks events

<details>
<summary>Tracks events (click to expand)</summary>

```
🔵 Tracked: support_opened <>
🔵 Tracked: support_chatbot_started <chat_id: chat UUID>
🔵 Tracked: support_new_request_failed <>
🔵 Tracked: support_chatbot_ticket_failure <chat_id: chat UUID, error_message: an error message>
```
</details>

### Error in chatbot web view

For this test, you'll need to checkout this branch and open the project in Xcode.

1. Checkout this branch in git
2. Download [this git patch](https://github.com/wordpress-mobile/WordPress-iOS/files/12424286/inspectable.patch)
3. Run `git apply inspectable.patch`
4. Run the project from Xcode
6. Tap "Contact support"
7. On the Mac, open Safari > Develop > choose the simulator
8. Open the Safari console and run `window.onerror("hello")`
9. Expect to see the following Tracks events (in the Xcode console or in the app logs):

```
🔵 Tracked: support_chatbot_webview_error <chat_id: chat UUID, error_message: hello>
```

## Regression Notes
1. Potential unintended areas of impact: None since we're just adding Tracks events
2. What I did to test those areas of impact (or what existing automated tests I relied on): I manually tested the Tracks events
3. What automated tests I added (or what prevented me from doing so): Since this PR only adds Tracks and we don't usually add tests for them, I decided to not add any tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist: No UI changes, so not applicable

Commit:
Add basic analytics events for chatbot

Commit:
Fix Zendesk identity creation flow in chat

When the user creates a Zendesk ticket via the chatbot, the app asks the user for their identity (email and name) if they haven't set it yet.
This flow wasn't working in the scenario where these weren't set. The fix involves passing in the view controller to use to present the identity form (an alert) on top of.

Commit:
Fix chatbot tests

guarani commented:
👋 Hey @staskus, I added the most important tracks events. The remaining ones are:
- `support_chatbot_webview_error`: This should be possible but I'm not sure we'll get errors from DocsBot (which is JavaScript loaded from a different origin)
- `support_chatbot_topic`: We're not explicitly segmenting technical questions from account/billing questions, so there's no way to get this
- `support_chatbot_user_submits_message`:  There's no API to get notified for every message the user writes and I don't think we need this event

I can add the error handing tomorrow, but I don't think the other events are worth adding (cc @irfano since you're working on the Android side of this). 

I'd also like to add Tracks events unit tests but will get to this tomorrow. Feel free to add to this PR if you'd like 🙇

staskus committed:
Merge branch 'trunk' into add/support-chatbot-analytics

staskus commented:
✅ I tested the scenarios and it works well!

> I can add the error handing tomorrow, but I don't think the other events are worth adding (cc @irfano since you're working on the Android side of this).

I agree that these events are not necessary.

> I'd also like to add Tracks events unit tests but will get to this tomorrow. 

Okay. I don't know if I unit-tested tracking of events myself before. I assume we'll need to mock `WPAnalytics.track(.` calls to be able to confirm whether they were called or not.

>  Feel free to add to this PR if you'd like 🙇

We first need to pass `chat_id, ticket_number` to already created events according to requirements. I'll start with that.

staskus committed:
Track chat_id parameter with support chatbot events

Generating a unique id in the scope of a chatbot session

staskus committed:
Track ticket_number when Zendesk ticket is created from chatbot

Parse ZDKRequest out of response from Zendesk's createRequest() call and pass to tracks properties

irfano commented on WordPress/Classes/ViewRelated/Support/SupportChatBot/SupportChatBotViewController.swift:
```diff
@@ -189,15 +189,16 @@ extension SupportChatBotViewController {
-                    self.viewModel.track(.supportChatbotTicketSuccess)
+                switch result {
+                case let .success(request):
+                    self.viewModel.track(.supportChatbotTicketSuccess, ticketNumber: request.requestId)
```
Have you tested what `request.requestId` returns? I'm not sure if it returns the correct id. I couldn't retrieve the actual ticket number, and I believe this is how it's intended in the Zendesk API. Since we're creating the request from the user side, the user can't be aware of the ticket number, as it's company-specific data.

guarani replied:
You're right, `requestId` is returning an alphanumeric ID instead of the numeric ticket ID we see in Zendesk. I'll keep looking.

guarani replied:
> I believe this is how it's intended in the Zendesk API. Since we're creating the request from the user side, the user can't be aware of the ticket number, as it's company-specific data.

That's a good point.

ravishanker commented:
We need to add this chat id in the bot configuration, to relate tracks with chat for reference like so

identify: {
customVariable: 'customValue', // This will be recorded in the question metadata and accessible via the API.
},

See https://github.com/wordpress-mobile/WordPress-Android/pull/19000/commits/ad6a9f5db6ab0e508f9ba26b0b9fe14dc6b457ea

Commit:
Merge branch 'trunk' into add/support-chatbot-analytics

Commit:
Added DocsBot chat ID to tracks

Commit:
Added web view error handler

Commit:
Prevent tracks event if user pulls down on chat screen

Commit:
Fix JS version issue

Commit:
Fix JS version issue

guarani commented:
👋 Thanks for the review folks! I've updated this to add the chat ID and the error messages in Tracks events.
This is ready for another review 🙇

ravishanker commented on WordPress/Classes/ViewRelated/Support/SupportChatBot/SupportChatBotViewController.swift:
```diff
@@ -58,11 +66,14 @@ final class SupportChatBotViewController: UIViewController {
-                id: '\(viewModel.id)',
-                 supportCallback: function (event, history) {
+                id: '\(viewModel.docsBotId)',
+                identify: {
```
👍

ravishanker reviewed (approved):
LGTM 👍
